compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -13266,6 +13266,268 @@ OMR::Z::TreeEvaluator::arraysetEvaluator(TR::Node * node, TR::CodeGenerator * cg
    return NULL;
    }
 
+
+void 
+OMR::Z::TreeEvaluator::genLoopForwardArrayCopy(TR::Node *node, TR::CodeGenerator *cg, TR::Register *byteSrcReg, TR::Register *byteDstReg, TR::Register *loopIterReg, bool isConstantLength)
+   {
+   TR::LabelSymbol *processingLoop = generateLabelSymbol(cg);
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, processingLoop);
+   generateSS1Instruction(cg, TR::InstOpCode::MVC, node, 255, 
+      new (cg->trHeapMemory()) TR::MemoryReference(byteDstReg, 0, cg), 
+      new (cg->trHeapMemory()) TR::MemoryReference(byteSrcReg, 0, cg));
+   generateRXInstruction(cg, TR::InstOpCode::LA, node, byteSrcReg, new (cg->trHeapMemory()) TR::MemoryReference(byteSrcReg, 256, cg));
+   generateRXInstruction(cg, TR::InstOpCode::LA, node, byteDstReg, new (cg->trHeapMemory()) TR::MemoryReference(byteDstReg, 256, cg));
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRCT, node, loopIterReg, processingLoop);
+   // If it is known at compile time that this is forward array copy and also length is known at compile time then internal control flow starts here.
+   if (node->isForwardArrayCopy() && isConstantLength)
+      processingLoop->setStartInternalControlFlow();
+   }
+
+
+void
+OMR::Z::TreeEvaluator::forwardArrayCopySequenceGenerator(TR::Node *node, TR::CodeGenerator *cg, TR::Register *byteSrcReg, TR::Register *byteDstReg, TR::Register *byteLenReg, TR::Node *byteLenNode, TR_S390ScratchRegisterManager *srm, TR::LabelSymbol *mergeLabel)
+   {
+#ifdef J9_PROJECT_SPECIFIC
+   bool mustGenerateOOLGuardedLoadPath = TR::Compiler->om.shouldGenerateReadBarriersForFieldLoads() &&
+                                         node->getArrayCopyElementType() == TR::Address;
+   if (mustGenerateOOLGuardedLoadPath)
+      {
+      // It might be possible that we have constant byte lenght load and it is forward array copy. 
+      // In this case if we need to do guarded Load then need to evaluate byteLenNode.
+      if (byteLenReg == NULL)
+         byteLenReg = cg->gprClobberEvaluate(byteLenNode);
+      TR::TreeEvaluator::genGuardedLoadOOL(node, cg, byteSrcReg, byteDstReg, byteLenReg, mergeLabel, srm, true);
```

Comment:
@fjeremic  This call here is implemented in https://github.com/eclipse/openj9/pull/3465. This is something on what I want your view, I moved the implementation as it handles special case for Concurrent Scavenge GC which really belongs to OpenJ9.

Reply:
I see two calls to `genGuardedLoadOOL`, one here and one in the backwards evaluator. I also see we call both of them in the generic path where we don't know if we are forwards or backwards. Won't this generate two OOL sequences for the fallback CS path when we only really need one?